extensions/coroutines-extensions/src/commonMain/kotlin/com/squareup/sqldelight/runtime/coroutines/FlowExtensions.kt

```diff
@@ -33,22 +33,18 @@ import kotlin.jvm.JvmName
 import kotlin.jvm.JvmOverloads
 
 /** Turns this [Query] into a [Flow] which emits whenever the underlying result set changes. */
+@ExperimentalCoroutinesApi
 @FlowPreview
 @JvmName("toFlow")
-fun <T : Any> Query<T>.asFlow(): Flow<Query<T>> = flow {
-  val channel = Channel<Unit>(CONFLATED)
+fun <T : Any> Query<T>.asFlow(): Flow<Query<T>> = flowViaChannel(CONFLATED) { channel ->
   val listener = object : Query.Listener {
     override fun queryResultsChanged() {
-      channel.offer(Unit)
+      channel.offer(this@asFlow)
     }
   }
   addListener(listener)
-  try {
-    emit(this@asFlow)
-    for (item in channel) {
-      emit(this@asFlow)
-    }
-  } finally {
+  channel.offer(this@asFlow)
+  channel.invokeOnClose {
```

Comment:
This needs to happen before the initial emission. If you apply `take(1)` you'll end up leaking the listener since it will cancel the lambda prior to the removal callback having been added.

Reply:
Fixed.